### PR TITLE
add support for rayserve in apiserver

### PR DESCRIPTION
**Pull request by scarlet25151**
## Why are these changes needed?
Same with ray job, we need to add support for CRUD of ray service in apiserver.

## Checks

- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scarlet25151** [comment]
For now, test on local environment has success and ready for review.

**Review by brucez-anyscale — APPROVED** [review]
thanks. lgtm. Please make sure the tests pass and let @Jeffwan approve

File: proto/serve.proto [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+      schema: {
+        json_schema: {
+          ref: ".api.Status";
+        }
+      }
+    }
+  }
+};
+
+service RayServeService {
```
Comment (Jeffwan) [inline proto/serve.proto:27]: not a revise suggestion: Please check Job side and make sure naming convention is exact same as Job.

File: proto/serve.proto [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  int32 total_size = 2;
+  // The token to list the next page of RayServices.
+  string next_page_token = 3;
+}
+
+message ListAllRayServicesRequest {
+  // A page token to request the next page of results. The token is acquried
+  // from the nextPageToken field of the response from the previous
+  // ListRayServices call or can be omitted when fetching the first page.
+  string page_token = 2;
```
Comment (Jeffwan) [inline proto/serve.proto:102]: should it start with 1?

File: proto/serve.proto [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  ActorOptions actor_options = 7;
+}
+
+message ActorOptions {
+  string runtime_env = 1;
+  double cpus = 2;
+  double gpu = 3;
+  int32 memory = 4;
+  int32 object_store_memory = 5;
+  string resource = 6;
```
Comment (Jeffwan) [inline proto/serve.proto:175]: what's the resource field for?

File: proto/serve.proto [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+message RayServiceStatus {
+  // The ray serve application status.
+  string application_status = 1;
+  // A human-readable description of the status of this operation.
+  string application_message = 2;
+  // The status for each deployment.
+  repeated ServeDeploymentStatus serve_deployment_status = 3;
+  // The related event for the ray service.
+  repeated RayServiceEvent ray_service_event = 4;
```
Comment (Jeffwan) [inline proto/serve.proto:187]: is this event from the serve endpoint?

File: apiserver/pkg/model/converter.go [hunk 5]
```diff
@@ -165,3 +165,113 @@ func FromCrdToApiJob(job *v1alpha1.RayJob) (pbJob *api.RayJob) {
...
+	apiServices := make([]*api.RayService, 0)
+	for _, service := range services {
+		apiServices = append(apiServices, FromCrdToApiService(service, serviceEventsMap[service.Name]))
+	}
+	return apiServices
+}
+
+func FromCrdToApiService(service *v1alpha1.RayService, events []v1.Event) *api.RayService {
+	defer func() {
+		err := recover()
```
Comment (Jeffwan) [inline apiserver/pkg/model/converter.go:179]: will we have any panic issue below?

File: apiserver/pkg/util/service.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+import (
+	"encoding/base64"
+
+	api "github.com/ray-project/kuberay/proto/go_client"
+	rayalphaapi "github.com/ray-project/kuberay/ray-operator/apis/ray/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const rayServiceDefaultVersion = "2.0"
```
Comment (Jeffwan) [inline apiserver/pkg/util/service.go:11]: 2.0.0?

**Comment by Jeffwan** [comment]
@scarlet25151 

1. Could you give a comparison between RayServe yaml and HTTP POST payload and that's easier to compare. 
2. Let's also make sure docs are updated. 
3. I do see few TODO items there, are they blockers for RayServe endpoint? If so, let's add the status summary in the PR description

File: proto/serve.proto (hunk 1 above)
Comment (scarlet25151) [reply proto/serve.proto:27]: yes, the terms here are exact aligned.

File: apiserver/pkg/model/converter.go (hunk 5 above)
Comment (scarlet25151) [reply apiserver/pkg/model/converter.go:179]: here is a check for the unexpected situation, I just make it aligned with the ray jobs https://github.com/ray-project/kuberay/blob/fdee883c3de7fa95fca6659bd200b98f2cc82b02/apiserver/pkg/model/converter.go#L130-L136

File: proto/serve.proto (hunk 2 above)
Comment (scarlet25151) [reply proto/serve.proto:102]: > @scarlet25151
> 
> 1. Could you give a comparison between RayServe yaml and HTTP POST payload and that's easier to compare.
yes, I will add it to the top of the conversation.
> 2. Let's also make sure docs are updated.
this part is not updated yet, I will update in the third commit.
> 3. I do see few TODO items there, are they blockers for RayServe endpoint? If so, let's add the status summary in the PR description
here is one things I would like to discuss, should we use the raycluster's ingress or rayservice ingress here? https://github.com/ray-project/kuberay/blob/fdee883c3de7fa95fca6659bd200b98f2cc82b02/ray-operator/controllers/ray/rayservice_controller.go#L601-L603

**Comment by scarlet25151** [comment]
@Jeffwan 

> 1. Could you give a comparison between RayServe yaml and HTTP POST payload and that's easier to compare.

yes, I will add it to the top of the conversation.

> 2. Let's also make sure docs are updated.

this part is not updated yet, I will update in the third commit.

> 3. I do see few TODO items there, are they blockers for RayServe endpoint? If so, let's add the status summary in 
the PR description

here is one things I would like to discuss, should we use the raycluster's ingress or rayservice ingress here? https://github.com/ray-project/kuberay/blob/fdee883c3de7fa95fca6659bd200b98f2cc82b02/ray-operator/controllers/ray/rayservice_controller.go#L601-L603

**Comment by scarlet25151** [comment]
The json body of the request would be like:
```json
{
  "name": "chenyu-test",
  "user": "chenyu.jiang",
  "serveDeploymentGraphSpec": {
      "importPath": "fruit.deployment_graph",
      "runtimeEnv": "working_dir: \"https://github.com/ray-project/test_dag/archive/c620251044717ace0a4c19d766d43c5099af8a77.zip\"",
      "serveConfigs": [{
        "deploymentName": "MangoStand",
        "replicas": 1,
        "userConfig": "price: 3",
        "actorOptions": {
          "cpus": 0.1
        }
      },
      {
        "deploymentName": "OrangeStand",
        "replicas": 1,
        "userConfig": "price: 2",
        "actorOptions": {
          "cpus": 0.1
        }
      },
      {
        "deploymentName": "PearStand",
        "replicas": 1,
        "userConfig": "price: 1",
        "actorOptions": {
          "cpus": 0.1
        }
      },
      {
        "deploymentName": "FruitMarket",
        "replicas": 1,
        "actorOptions": {
          "cpus": 0.1
        }
      },{
        "deploymentName": "DAGDriver",
        "replicas": 1,
        "routePrefix": "/",
        "actorOptions": {
          "cpus": 0.1
        }
      }]
  },
  "clusterSpec": {
  // ...
  }
}
```
which is compatible to the example of the yaml:
https://github.com/ray-project/kuberay/blob/40ea21ccac8d1672633f27952e485d31aee8fb0e/ray-operator/config/samples/ray_v1alpha1_rayservice.yaml#L11-L43

File: proto/serve.proto (hunk 2 above)
Comment (scarlet25151) [reply proto/serve.proto]: done.

File: proto/serve.proto (hunk 3 above)
Comment (Jeffwan) [reply proto/serve.proto:175]: Minor: This is actually the custom resource? I feel ray serve's definition is not that clear. As a user, I may be confused on cpu,memory,gpu and resources here. I can not figure out what to put. Do you have the same feeling? If we can use different name to clarify that, that would be great.

File: proto/serve.proto (hunk 4 above)
Comment (Jeffwan) [reply proto/serve.proto:187]: sounds good. BTW, let's name it `ray_service_events`

File: apiserver/pkg/model/converter.go (hunk 5 above)
Comment (Jeffwan) [reply apiserver/pkg/model/converter.go:176]: sounds good

**Comment by Jeffwan** [comment]
> here is one things I would like to discuss, should we use the raycluster's ingress or rayservice ingress here?

Do you think rayserve should leverage raycluster's ingress to accept external traffic?

**Comment by scarlet25151** [comment]
> > here is one things I would like to discuss, should we use the raycluster's ingress or rayservice ingress here?
> 
> Do you think rayserve should leverage raycluster's ingress to accept external traffic?

I see, here is an outdate comment, for now I've implement in the way to poulate response with active rayservice cluster's ingress and endpoints .

File: proto/serve.proto (hunk 3 above)
Comment (scarlet25151) [reply proto/serve.proto]: yes, I've replace the name with `cpu_per_actor` ... and add more comments for this field.

File: proto/serve.proto (hunk 4 above)
Comment (scarlet25151) [reply proto/serve.proto]: Done.

**Review by Jeffwan — APPROVED** [review]

**Pull request merged**